citizennet/purescript-halogen-select
# Free query

## What does this pull request do?



Base the Select's Query off a Free Monad approach. The old `Query` is now called `QueryF`, and `Query` is now `Free (QueryF ...)`. This allows using all of the monadic language to express what should happen when events trigger now.

As an example of putting this to use, I changed the logic in `onMouseDown` in `setToggleProps` to get the current visibility of the element and then choose to blur or focus it based on the current visibility, as well as toggling the visibility. (Now toggling is no longer a primitive!)

I also added `setItemPropsAndBlur` to allow blurring after selecting an item. This behavior was possible before – I think it might be a little more intuitive, but something we can discuss.

There is a little more boilerplate in the `Select` module, to help with the free queries (mostly just adding `liftF` and applying `Unit` by default), but this helps a _tiny_ bit with the usual Halogen boilerplate (`Just <<<` instead of `H.action` I guess). `Select.always` helps with the remaining boilerplate. Note how `toggleVisibility` is now implemented via `getVisibility` and `setVisibility`.

## Where should the reviewer start?



Implementation changes:
- for `Query`: see `toggleVisibility`, `triggerFocus`, and `eval`

Behavior changes:
- As noted above, see `setToggleProps` and `setItemPropsAndBlur`

Otherwise it should be the same code expressed differently ;)

## How should this be manually tested?



In the documentation site, observe that:
1. Clicking on an item (in either the typeahead or dropdown) will blur the input element.
2. Clicking on the dropdown toggle when it is already open will blur the toggle element.

## Other Notes:


Also added some tips to the readme for building the documentation site. YMMV.
